### ISSUE TITLE
The vscode extension should be cleaned up and released

### ISSUE THREAD

**Issue by thoughtpolice** [facebook/buck2#309]
I think it would be nice if the vscode extension was released and polished and usable for Buck2. I have it installed locally, but it's a chore especially when using features like remote vscode.

What would be needed for this? Ideally it would be released through GitHub as well, which would be really nice to help people contribute.

A tricky thing is how much buck2-specific "stuff" should be in it. The extension is distributed with the starlark package, but right now buck2 is the main user. Ideally, buck2 *specific* integration would include things like running or listing build targets...

**Comment by ndmitchell** [comment]
Internally this is not very well loved or maintained. I wrote the first version. Then someone who is no longer with the company picked it up, did a little bit to it, but it is basically incomplete as things stand. Agreed it would be great to be cleaned up.

I think the biggest question is whether we should have a Buck2 extension, or a separate Starlark one. My guess is we should have a separate Buck2 extension, especially for features like DAP/debugger support which will require something quite different to normal Starlark. CC @cjhopman

**Comment by thoughtpolice** [comment]
I agree there should be a Buck2 specific extension, but if things are unloved as they are &mdash; maybe the existing version should just be forked into this repository and a Buck2-specific extension should be what we focus on.

Like, yes, in some better world would be nice to write a really good generalized Starlark extension, then build off that. But if it's mostly unloved, I think probably focusing on actual changes that bring users visible value would be a better use of time, for now.

**Comment by ndmitchell** [comment]
I think there's a separation between the extension (the bit of TypeScript that wires it up) and the LSP server (the thing that does most of the work). For Buck2, most of the effort is likely to be in the TypeScript, but the underlying LSP should mostly work from starlark-rust with slightly different instantiation. So as long as the Rust code goes in the right place, focusing only on the Buck2 TypeScript bits makes sense.

**Comment by thoughtpolice** [comment]
Right, so to fully clarify for anyone reading, there're three parts here:

- Rust: The `starlark` package which has all the fundamental nuts and bolts (LSP, DAP, and the ability to define built in APIs)
- Rust: The buck2 LSP implementation; basically, this is just re-using previous rust code as a library, but since it knows about buck2 APIs, BUCK files, etc, it's designed for our APIs. The `buck2 lsp` command allows users to launch the LSP on a project, and
- Typescript: The actual Visual Studio extension, which runs in the editor, which talks to the LSP and tells the editor what targets are available and communicates breakpoints, etc etc. It also will need some Buck2 specific knowledge.

Every user of the starlark package gets the code in point 1 for free. But point 2 and 3 must be done on a case-by-case basis for each project. (Point 2 is obviously always required even without vscode support, keep in mind.) While it would be nice to *also* share a lot of code for point 3 to the greatest extent possible, I instead suggest we just make a Visual Studio Code buck2 package, specifically for our uses, and focus our efforts on that, yes.

Maybe at some point in the future, some other user of the starlark package might want to create their own vscode extension, and share some code between theirs and the buck2 vscode extension. We could cross that bridge when we come to it.

**Comment by cjhopman** [comment]
> My guess is we should have a separate Buck2 extension, especially for features like DAP/debugger support which will require something quite different to normal Starlark. CC @cjhopman

Internally we have a fork of the starlark extension that both customizes it for buck2 and for some internal things, but we still call it "starlark". I've been leaning towards we should just commit to that being a buck2 extension, otherwise we'll find ourselves needing to sort of arbitrarily need to put things in that extension or another depending on how buck2 specific it is. I think that would also make it easier to have that extension share code with a public buck2 extension (i looked into getting it to share more with the starlark one and the internal changes were sort of easy to handle but actually the buck2 customizations were harder). That probably makes more sense also if there's some place where people want to use a different embedding of starlark as well (i mean, can you just imagine some game studio using both buck2 for builds and an embedded starlark for their scripting language, how wonderful!).
